### ISSUE TITLE
GraphStoreProtocol fails with HTTPS credentials

### ISSUE THREAD

**Issue by rizktouma** [linkedpipes/etl#685]
I am trying to use the GraphStoreProtocol plugin to upload data to a Virtuoso HTTPS endpoint. The component fails execution and returns a `Can't upload data, status: 401` exception.

I took a look at the source code of the plugin, specifically the `executeHttp()` method, and I tried the following solutions:

- Using AuthScope.ANY when creating the CredentialsProvider (to make sure that it is adding the credentials for HTTP as well as HTTPS):

```java
final CredentialsProvider creds = new BasicCredentialsProvider();
creds.setCredentials(AuthScope.ANY, new UsernamePasswordCredentials(
         configuration.getUserName(), configuration.getPassword()));
```

- Adding an AuthCache and the same CredentialsProvider to the created context:

```java
AuthCache authCache = new BasicAuthCache();
authCache.put(new HttpHost(configuration.getEndpoint()), new BasicScheme());
context.setCredentialsProvider(creds);
context.setAuthCache(authCache);
```

- Creating the Authorization header and adding it manually to `httpMethod` before executing it with `httpClient`:

```java
String userPwStr = configuration.getUserName() + ":" + configuration.getPassword();
httpMethod.addHeader("Authorization", "Basic " +
         Base64.getEncoder().encodeToString(userPwStr.getBytes()));
```

Only the last solution worked, even without adding the AuthCache to the context. So it strangely seems that the default CredentialsProvider of the HttpClient is not adding the Authorization header to the requests.

I tried to look for possible causes online but couldn't find anything. Could you take a look at this issue?

Cheers,
Rizkallah

**Comment by skodapetr** [comment]
@rizktouma Thanks for the report, can you share some details about how is the endpoint setup? It would help us reproduce and examine the issue.

**Comment by jakubklimek** [comment]
@rizktouma Could you provide the exact web server settings?

Virtuoso uses Digest authentication method, not Basic authentication method. In our setting, there is a Virtuoso HTTP endpoint with Virtuoso's Digest authentication, proxied by nginx as HTTPS (with no additional authentication).

It might be the case that you do some Basic authentication outside of Virtuoso (e.g. in a proxy), which will currently not work.

**Comment by rizktouma** [comment]
Apologies for the late reply. I am using an external server and I do not have control over the server settings so I am not quiet sure what you mean by "server settings".

In any case, due to the requirements of the project we changed to GraphDB which is not officially supported by LinkedPipes. So I am currently working on an extension of the plugin to connect to GraphDB as well. So far it seems to be working when manually creating the Basic Authorization HTTP header, but not with the CredentialsProvider.

**Comment by rizktouma** [comment]
A quick update: I created a new plugin to store RDF data into GraphDB with the Graph Store Protocol and I had to specify the Authorization header manually for it to work correctly. I tested it on a local repository and a remote one and I was getting the same error as above in both cases.

That was actually the only modification I needed for the plugin to work correctly with a GraphDB repository. Otherwise, the graph store protocol of GraphDB expects the same URI parameters and structure as Virtuoso.

I'm going to close this issue since I am not sure if this was a problem only I'm having locally.
